### PR TITLE
Read the list of repositories for the `catalog` command from the terragrunt configuration file

### LINKED ISSUE #2828: Support for loading the catalog from one or more repos
**Describe the solution you'd like**

We should add support for loading the `catalog` from one or more repositories defined in the `terragrunt.hcl` file.
This way, users could use this `terragrunt.hcl` as a "config file" where they define all the repos that contain their catalog in an infra-live repo, and if in that repo (which is where you'd be doing scaffolding anyway), they could run `terragrunt catalog` with no other args, and have the proper catalog loaded automatically.

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

`terragrunt catalog <repo-url>`

If `<repo-url>` is provided, the repository will be cloned into a temporary directory, otherwise:

1. The repository list are searched in the config file `terragrunt.hcl`. if `terragrunt.hcl` does not exist in the current directory, the config are searched in the parent directories.
1. If the repository list is not found in the configuration file, the modules are looked for in the current directory.

An example of how to define the list of repositories for the `catalog` command in the `terragrunt.hcl` configuration file:

``` hcl
catalog {
  urls = [
    "relative/path/to/repo", # will be converted to the absolute path, relative to the path of the configuration file.
    "/absolute/path/to/repo",
    "github.com/gruntwork-io/terraform-aws-lambda", # url to remote repository
    "http://github.com/gruntwork-io/terraform-aws-lambda", # same as above
  ]
}
```

Fixes #2828

### CODE REVIEW

File: config/catalog.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	log.Debugf("Reading Terragrunt cfg file at %s", configPath)
+
+	configString, err := util.ReadFileAsString(configPath)
+	if err != nil {
+		return err
+	}
+
+	// Parse the HCL string into an AST body that can be decoded multiple times later without having to re-parse
+	parser := hclparse.NewParser()
+	hclFile, err := parseHcl(parser, configString, configPath)
```
Comment (brikis98) [inline config/catalog.go:34]: It looks like you're doing custom parsing of the TG config here... Why not using the existing code that already knows how to do it?

One issue here is that the TG config file supports all sorts of features, such as `locals`, functions, etc... And if any of those are used in the `catalog` block, it's not clear if this custom parsing will support them.

**Review by brikis98 — COMMENTED** [review]
Very nice!

Just one question on config parsing.

**Review by denis256 — COMMENTED** [review]
Looks like `catalog` block from parent include is not handled

```
# terragrunt.hcl
include {
  path = find_in_parent_folders()
}

# ../terragrunt.hcl
catalog {
  urls = [
    "github.com/gruntwork-io/terraform-aws-lambda",
  ]
}


$ terragrunt catalog
ERRO[0000] the specified path "/projects/gruntwork/terragrunt-tests/pr-2852/app" is not a git repository 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
```

if I copy parent terragrunt.hcl, things start to work:
```
$ cp ../terragrunt.hcl .
$ terragrunt catalog
INFO[0000] Cloning repository "git::https://github.com/gruntwork-io/terraform-aws-lambda.git" to temprory directory "/tmp/catalog-31484d765750376878357a6b595274367356677931475962474a67"
```

https://github.com/denis256/terragrunt-tests/tree/master/pr-2852

File: config/catalog.go (hunk 1 above)
Comment (levkohimins) [reply config/catalog.go]: Ah, you are right. Redone. Thanks!

**Comment by levkohimins** [comment]
> Looks like `catalog` block from parent include is not handled
> 
> ```
> # terragrunt.hcl
> include {
>   path = find_in_parent_folders()
> }
> 
> # ../terragrunt.hcl
> catalog {
>   urls = [
>     "github.com/gruntwork-io/terraform-aws-lambda",
>   ]
> }
> 
> 
> $ terragrunt catalog
> ERRO[0000] the specified path "/projects/gruntwork/terragrunt-tests/pr-2852/app" is not a git repository 
> ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1 
> ```
> 
> if I copy parent terragrunt.hcl, things start to work:
> 
> ```
> $ cp ../terragrunt.hcl .
> $ terragrunt catalog
> INFO[0000] Cloning repository "git::https://github.com/gruntwork-io/terraform-aws-lambda.git" to temprory directory "/tmp/catalog-31484d765750376878357a6b595274367356677931475962474a67"
> ```
> 
> https://github.com/denis256/terragrunt-tests/tree/master/pr-2852

I redone the code. Now it's fixed.

File: util/file.go [hunk 2]
```diff
@@ -627,3 +628,25 @@ func IsDirectoryEmpty(dirPath string) (bool, error) {
 	}
 	return true, nil
 }
+
+// FindFileInParentDirs returns `path` if the file exists, otherwise searches for a file with the same name in the parent directories and returns the file's path if found.
+func FindFileInParentDirs(path string) string {
```
Comment (brikis98) [inline util/file.go:633]: Sorry, should've mentioned it: https://github.com/gruntwork-io/terragrunt/blob/master/config/config_helpers.go#L410

**Review by brikis98 — COMMENTED** [review]
Woot, very nice!

Just one final item to look at.

File: util/file.go (hunk 2 above)
Comment (levkohimins) [reply util/file.go]: Oh, no, it's my fault. I should have known about this feature. Thanks for the advice! Redone.

**Review by brikis98 — APPROVED** [review]
LGTM!

**Pull request merged**